src/main/java/frc/robot/utils/FieldFinder.java

```diff
@@ -0,0 +1,207 @@
+            double x, double y, double dx, double dy, FieldLocations location) {
+        switch (location) {
+            case RED_STAGE:
+                return willIHitTriangle(
+                        x,
+                        y,
+                        dx,
+                        dy,
+                        FieldLocations2024.STAGE_RED_NEAR_SIDE_X,
+                        FieldLocations2024.STAGE_RED_NEAR_SIDE_Y,
+                        FieldLocations2024.STAGE_RED_LEFT_SIDE_X,
+                        FieldLocations2024.STAGE_RED_LEFT_SIDE_Y,
+                        FieldLocations2024.STAGE_RED_RIGHT_SIDE_X,
+                        FieldLocations2024.STAGE_RED_RIGHT_SIDE_Y);
+            case BLUE_STAGE:
+                return willIHitTriangle(
+                        x,
+                        y,
+                        dx,
+                        dy,
+                        FieldLocations2024.STAGE_BLUE_NEAR_SIDE_X,
+                        FieldLocations2024.STAGE_BLUE_NEAR_SIDE_Y,
+                        FieldLocations2024.STAGE_BLUE_LEFT_SIDE_X,
+                        FieldLocations2024.STAGE_BLUE_LEFT_SIDE_Y,
+                        FieldLocations2024.STAGE_BLUE_RIGHT_SIDE_X,
+                        FieldLocations2024.STAGE_BLUE_RIGHT_SIDE_Y);
+            case RED_WING:
+                return x > FieldLocations2024.WING_RED_END_X
+                        || x + dx > FieldLocations2024.WING_RED_END_X;
+            case BLUE_WING:
+                return x < FieldLocations2024.WING_BLUE_END_X
+                        || x + dx < FieldLocations2024.WING_BLUE_END_X;
+            case MIDDLE:
+                return (x < FieldLocations2024.WING_RED_END_X
+                                || x + dx < FieldLocations2024.WING_RED_END_X)
+                        && (x > FieldLocations2024.WING_BLUE_END_X
+                                || x + dx > FieldLocations2024.WING_BLUE_END_X);
+            default:
+                return false;
+        }
+    }
+
+    private static boolean willIHitTriangle(
+            double x,
+            double y,
+            double dx,
+            double dy,
+            double x1,
+            double y1,
+            double x2,
+            double y2,
+            double x3,
+            double y3) {
+        double m = dy / dx;
+
+        double m1 = (y2 - y1) / (x2 - x1);
+        double m2 = (y3 - y2) / (x3 - x2);
+        double m3 = (y1 - y3) / (x1 - x3);
+
+        double b = y - m * x;
+
+        double b1 = y1 - m1 * x1;
+        double b2 = y2 - m2 * x2;
+        double b3 = y3 - m3 * x3;
+
+        double xIntercept1 = (b1 - b) / (m - m1);
+        double xIntercept2 = (b2 - b) / (m - m2);
+        double xIntercept3 = (b3 - b) / (m - m3);
+
+        double yIntercept1 = (b / m - b1 / m1) / (1 / m - 1 / m1);
+        double yIntercept2 = (b / m - b2 / m2) / (1 / m - 1 / m2);
+        double yIntercept3 = (b / m - b3 / m3) / (1 / m - 1 / m3);
```

Comment:
We need to add divide by 0 protection in this area. Also if dx=0, it seems like this would fail? 

However, I just looked, and I think we should use Java's built-in line intersection logic: [StackOverflow Link](https://stackoverflow.com/questions/16333650/how-to-check-whether-2-lines-segments-intersect). This would really simplify this function (since we would just be checking for 3 line intersections and throwing some OR gates at it)

Reply:
That's a lot easier...